### PR TITLE
[bsp][synopsys] add basic new embarc bsp support

### PR DESCRIPTION
**Pull request by IRISZZW**
#3125  拉取/合并请求描述：(PR description)

[
* the initial support of synopsys designware ARC processor
  using embARC_BSP, all synopsys ARC-based boards are
  supported:
  -ARC Software Development Platform
  -ARC EM Starter Kit
  -ARC EM Software Development Platform
  -ARC HS Development Kit
  -ARC IoT Development Kit

* The embARC BSP is a new generation embARC software development
  package.  ​It is designed to be the inter-layer between hardware and
  operating system. ​ BSP could hide the difference of hardware/boards,
  provide a unified interface to upper-layer.

* the initial support of synopsys MWDT toolchain.
  The DesignWare® ARC® MetaWare Development Toolkit builds upon
  a 25-year legacy of industry-leading compiler and debugger products.
  It is a complete solution that contains all the components needed to
  support the development, debugging and tuning of embedded applications
  for the DesignWare ARC processors.

* for detailed board information, pls go embarc.org.
]

以下的内容不应该在提交PR时的message修改，修改下述message，PR会被直接关闭。请在提交PR后，浏览器查看PR并对以下检查项逐项check，没问题后逐条在页面上打钩。
The following content must not be changed in submitted PR message. Otherwise, the PR will be closed immediately. After submitted PR, please use web browser to visit PR, and check items one by one, and ticked them if no problem.

### 当前拉取/合并请求的状态 Intent for your PR

必须选择一项 Choose one (Mandatory):

- [ ] 本拉取/合并请求是一个草稿版本 This PR is for a code-review and is intended to get feedback
- [x] 本拉取/合并请求是一个成熟版本 This PR is mature, and ready to be integrated into the repo

### 代码质量 Code Quality：

我在这个拉取/合并请求中已经考虑了 As part of this pull request, I've considered the following:

- [x] 已经仔细查看过代码改动的对比 Already check the difference between PR and old code
- [x] 代码风格正确，包括缩进空格，命名及其他风格 Style guide is adhered to, including spacing, naming and other style
- [x] 没有垃圾代码，代码尽量精简，不包含`#if 0`代码，不包含已经被注释了的代码 All redundant code is removed and cleaned up
- [x] 所有变更均有原因及合理的，并且不会影响到其他软件组件代码或BSP All modifications are justified and not affect other components or BSP
- [x] 对难懂代码均提供对应的注释 I've commented appropriately where code is tricky
- [x] 本拉取/合并请求代码是高质量的 Code in this PR is of high quality

### CODE REVIEW

File: libcpu/arc/em/contex_gcc_mw.S [hunk 1]
```diff
@@ -4,20 +4,42 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 #define __ASSEMBLY__
```
Comment (BernardXiong) [inline libcpu/arc/em/contex_gcc_mw.S]: contex_gcc_mw.S, which used for GCC based toolchain?

File: libcpu/arc/em/cpuport.c [hunk 2]
```diff
@@ -66,4 +69,19 @@ rt_uint8_t *rt_hw_stack_init(void       *tentry,
 void rt_hw_exception_install(rt_err_t (*exception_handle)(void *context))
 {
     exception_handle = exception_handle;
-}
\ No newline at end of file
+}
+
+void vPortSetStackCheck(rt_uint32_t *from, rt_uint32_t *to)
```
Comment (BernardXiong) [inline libcpu/arc/em/cpuport.c]: Please provide a more suitable name.

File: libcpu/arc/em/contex_gcc_mw.S (hunk 1 above)
Comment (IRISZZW) [reply libcpu/arc/em/contex_gcc_mw.S]: file contex_gcc_mw.S is used for gcc toolchain and metaware(mw) toolchain both.

File: libcpu/arc/em/cpuport.c (hunk 2 above)
Comment (IRISZZW) [reply libcpu/arc/em/cpuport.c]: already done.

**Pull request merged**